department-of-veterans-affairs/vets-website
# Aedp/aw/vadx dev server mgmt

## Are you removing, renaming or moving a folder in this PR?

Did you change site-wide styles, platform utilities or other infrastructure?

## Summary

- In an effort to make local development easier, I have created a 'server that runs your servers'
- The main idea is that it runs the local FE dev server and mock api server at the same time, but within two different processes.
- This allows a single command to be used to start a development environment
- Includes a 'management' server that is opened up locally on port `1337`
- Endpoints on the management server include
  -  `/events` a SSE (server sent event) endpoint so a frontend can subscribe to change events emitted by the management server
  - `/manifests` get a list of all the applications that _could_ be run locally. This gets all the entryNames present so you could watch any combination of apps if you so desire
  - `/output` get an output cache of a process name. In a nutshell this gets all the terminal output history for an existing process
  - `/start-frontend-server` start the frontend dev server with a set of entry names to run the build with
  - `/start-mock-server` start mock api with a response path to use as source of mock responses
  - '/status` get the status of any running processes managed by the server
  - `/stop-on-port` stops a process on a port, validates port range beforehand
- Adds a package.json to the `_mock-form-ae-design-patterns` so that our application can be used as a Yarn workspace and have it's own scripts / dependencies


## Related issue(s)

- https://github.com/department-of-veterans-affairs/tmf-auth-exp-design-patterns/issues/257

## Testing done

- Steps to test:
  - run a fresh `yarn install` on the repo to make sure yarn workspaces are correct and aggregated
  - cd 'src/applications/_mock-form-ae-design-patterns`
  - run `yarn vadx` the process should output something like `webpack compiled successfully in 15297 ms` when it is finished building
  - verify that the FE webpack dev server and mock-api both start in the same terminal tab. If you scroll up the terminal feed,  you should see various outputs for each process `[fe-dev-server]` and `[mock-server]`
  - load `http://localhost:3001/mock-form-ae-design-patterns` and verify that the application works as expected

- Tested locally via Yarn workspace
- Used postman / http client to test server poc endpoints

## Screenshots

No UI changes, just developer tooling additions

## What areas of the site does it impact?

local development only

## Acceptance criteria

- yarn command works to spool up the development environment

### Quality Assurance & Testing


### Error Handling


### Authentication
